In [245]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import r2_score
import numpy as np

## Part 1. Loading the dataset

In [246]:
# Using pandas load the dataset (load remotely, not locally)
url = "/workspaces/assignment-4-equation-of-a-slime-<redacted>/science_data_large.csv"
df = pd.read_csv(url)

# Output the first 15 rows of the data
print("First 15 rows of the dataset:")
print(df.head(15))

# Display a summary of the table information (number of datapoints, etc.)
print("\nDataset Summary:")
print(df.info())

num_data_points = df.shape[0]
print(f"Number of data points: {num_data_points}")

First 15 rows of the dataset:
    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

Dataset Summary:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        100

## Part 2. Splitting the dataset

In [247]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = df[['Temperature °C','Mols KCL']]  # Features: KCl and Temperature
y = df['Size nm^3']           # Label: Size_Change
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## Part 3. Perform a Linear Regression

In [248]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_point = np.array([[25, 5]])  # Example: 5 KCl, 25 Temperature, temp is 470
prediction = model.predict(sample_point)
print(f"Prediction for sample point {sample_point[0]}: {prediction[0]}")

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
#train_score = model.score(X_train, y_train)
test_score = model.score(X_test, y_test)

#print(f"Training set score: {train_score}")
print(f"Test set score: {test_score}")

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_

print(f"Coefficients: {coefficients}")
print(f"Intercept: {intercept}")

Prediction for sample point [25  5]: -382574.34391668066
Test set score: 0.8552472077276095
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340764


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


## Model Score Explanation

The score reported by the model is the coefficient of determination (R-squared). 
For the test set, we got a score of 0.8552472077276095.

This score represents the proportion of the variance in the dependent variable (Size_Change) 
that is predictable from the independent variables (KCl and Temperature). 
A score of 1.0 indicates perfect prediction, while a score of 0 indicates that the model 
is just predicting the mean of the target variable.

Our score indicate that the model performs well on the test set. This suggests that our linear regression model has captured the underlying relationship between the input variables (KCl and Temperature) and the output variable (Size_Change) effectively. The model explains a large portion of the variance in the data and is likely to make good predictions on new, unseen data.

## Equation of the Model

The linear regression model can be represented by the following equation:

$$h(x) = {coefficients[0]}x_1 + {coefficients[1]}x_2 + {intercept}$$
$$h(x) = {866.15}x_1 + {1032.7}x_2 - {409390}$$


Where:
- $x_1$ represents the Temperature in °C
- $x_2$ represents the KCl concentration in Mols
- $h(x)$ is the predicted Size_Change

This equation allows us to predict the Size_Change for any given values of KCl and Temperature.

Sample equation: $E = mc^2$

## Part 4. Use Cross Validation

In [249]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data

cv_scores = cross_val_score(model, X, y, cv=10, scoring='r2')

# Print individual scores
print("Cross-validation scores:", cv_scores)

# Print mean and standard deviation of the scores
print("Mean R2 score:", np.mean(cv_scores))
print("Standard deviation of R2 score:", np.std(cv_scores))

# Report on their finding and their significance
# In Markdown

Cross-validation scores: [0.81123596 0.86440978 0.87808742 0.86561069 0.87495621 0.84484397
 0.87941022 0.86349411 0.78353682 0.88686516]
Mean R2 score: 0.8552450341984701
Standard deviation of R2 score: 0.03152876296534241


## Cross-Validation Results

We performed 5-fold cross-validation on our linear regression model. Here are the results:

Cross-validation scores: [0.81123596 0.86440978 0.87808742 0.86561069 0.87495621 0.84484397
 0.87941022 0.86349411 0.78353682 0.88686516]
Mean R2 score: 0.8552450341984701
Standard deviation of R2 score: 0.03152876296534241

### Interpretation of Results

1. **Consistency of Performance**: 
   The range of R2 scores across different folds indicates how consistently our model performs on different subsets of the data. A small range suggests consistent performance, while a large range might indicate that our model's performance is sensitive to the specific data it's trained on.

2. **Average Performance**: 
   The mean R2 score gives us an estimate of how well our model performs on average. This is likely a more reliable indicator of the model's true performance than the single score we got from our initial train-test split. The high mean \( R^2 \) value (0.8552) indicates that the model explains a significant portion of the variability in the target variable, indicating good predictive performance.

3. **Variability in Performance**: 
   The standard deviation of the R2 scores tells us about the variability in our model's performance. A low standard deviation suggests that our model's performance is stable across different subsets of the data.
   The relatively low standard deviation (0.0315) suggests that the model's performance is consistent, which is crucial for reliability when deploying the model in real-world scenarios.

4. **Model Reliability**: 
   If our cross-validation scores are consistently high with low variability, it suggests that our linear regression model is capturing a genuine relationship in the data that generalizes well across different subsets.

5. **Potential Overfitting**: 
   If we see high variability in the scores or significantly lower scores compared to our initial results, it might indicate that our model is overfitting to specific characteristics of the training data.

### Significance

These cross-validation results provide a robust evaluation of our model's performance. They give us high confidence in the ability of our linear regression model to predict Size_Change based on KCl concentration and Temperature. 

The small variability in performance suggests that the model's predictions are likely to be consistently reliable across different samples from the same population.

These results indicate that the linear relationship we've modeled between KCl concentration, Temperature, and Size_Change is strong and consistent. This suggests that our equation derived from the linear regression coefficients is likely to be a good representation of the underlying relationship in the data.

Given these strong results, we can proceed with confidence in using this model for predicting Size_Change based on KCl concentration and Temperature. 

## Part 5. Using Polynomial Regression

In [250]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
X_poly_train = poly.fit_transform(X_train)
X_poly_test = poly.transform(X_test)

# Train the model
model_poly = LinearRegression()
model_poly.fit(X_poly_train, y_train)

# Make predictions
y_pred_test = model_poly.predict(X_poly_test)

# Report on the metrics and output the resultant equation as you did in Part 3.
# Calculate R2 scores
r2_test = model_poly.score(X_poly_test, y_test)

#print(f"Training R2 score: {r2_train}")
print(f"Test R2 score: {r2_test}")

coefficients = model_poly.coef_
intercept = model_poly.intercept_

print(f"Coefficients: {coefficients}")
print(f"Intercept: {intercept}")

Test R2 score: 1.0
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.27195488e-07  1.26494371e-11
  2.00000000e+00  2.85714287e-02]
Intercept: 2.0477105863392353e-05



## Polynomial Regression Results (Degree 2)

### Process
1. We used PolynomialFeatures to transform our original features into polynomial features of degree 2.
2. We applied this transformation to both our training and test sets.
3. We then trained a linear regression model on these polynomial features.
4. Finally, we made predictions on the test set and calculated the R2 score.

### Metrics
- Test R2 score: 1.0

This R2 score indicates how well our polynomial model fits the test data. An R2 score closer to 1 indicates a better fit. This score is higher than the linear model's score from Part 3, suggesting better performance.

### Equation
The polynomial regression model of degree 2 can be represented by the following equation:

$$h(x) = a_0 + a_1x_1 + a_2x_2 + a_3x_1^2 + a_4x_1x_2 + a_5x_2^2$$

Where:
- $x_1$ represents the KCl concentration
- $x_2$ represents the Temperature
- $h(x)$ is the predicted Size_Change

Substituting the coefficients:

$$h(x) = 2.04771[10^{-5}] + [1.2][10^{1}]x_1 - [1.27195][10^{-7}]x_2 + [1.26494][10^{-11}]x_1^2 + [2.0]x_1x_2 + [2.85714][10^{-2}]x_2^2$$

### Interpretation

1. Constant term ([2.047710]): This is the predicted Size_Change when both KCl and Temperature are zero.
2. Linear terms ($x_1$ and $x_2$): These represent the linear effects of KCl and Temperature on Size_Change.
3. Quadratic terms ($x_1^2$ and $x_2^2$): These capture any non-linear relationships between each variable and Size_Change.
4. Interaction term ($x_1x_2$): This represents how the effect of KCl on Size_Change depends on the Temperature, and vice versa.

The magnitude and sign of each coefficient tell us about the strength and direction of each effect:
- Positive coefficients indicate that increasing that term increases Size_Change.
- Negative coefficients indicate that increasing that term decreases Size_Change.
- Larger absolute values indicate stronger effects.


### Comparison to Linear Model

In conclusion, the higher R2 score does suggest better performance of the polynomial model. This improvement indicates that the relationship between KCl concentration, Temperature, and Size_Change is not strictly linear, and the polynomial terms are helping to capture more complex patterns in the data. However, always balance this improved performance against the increased complexity of the model and ensure that the improvement is meaningful in the context of your specific problem.

This polynomial model allows us to capture more complex relationships between KCl concentration, Temperature, and Size_Change, potentially improving our predictions compared to the linear model. However, it's important to balance this improved fit against the increased complexity and potential for overfitting.